### PR TITLE
Timeseries converter

### PR DESCRIPTION
**Pull request by swillner**
- [x] Implementation finished
- [x] Existing tests adjusted
- [x] Missing tests added
- [x] Documentation adjusted
- [x] Notebooks adjusted
- [x] Description in ``CHANGELOG.rst`` added (single line such as: ``(`#XX <https://github.com/openclimatedata/openscm/pull/XX>`_) Added feature which does something``)

### CODE REVIEW

File: openscm/core.py [hunk 1]
```diff
@@ -256,19 +275,27 @@ def get_writable_timeseries_view(
...
         """
         Get a parameter or ``None`` if not found.
 
         Parameters
         ----------
         name
             :ref:`Hierarchical name <parameter-hierarchy>` of the parameter
-        region
+        region_name
             Hierarchical name of the region or ``()`` for "World".
```
Comment (znicholls) [inline openscm/core.py]:
```suggestion
            Hierarchical name of the region".
```

I thought we got rid of this?

File: openscm/timeseries_converter.py [hunk 2]
```diff
@@ -0,0 +1,364 @@
+"""
+Different climate models often use different time frames for their input and output
+data. This includes different 'meanings' of time steps (e.g. beginning vs middle of year) and
+different lengths of the time steps (e.g. years vs months). Accordingly, OpenSCM
+supports the conversion of timeseries data between such timeseriess, which is handled in
```
Comment (znicholls) [inline openscm/timeseries_converter.py]:
```suggestion
supports the conversion of timeseries data between such timeseries, which is handled in
```

File: openscm/timeseries_converter.py [hunk 3]
```diff
@@ -0,0 +1,364 @@
...
+this module. A thorough explaination of the procedure used is given in a dedicated
+`Jupyter Notebook
+<https://github.com/openclimatedata/openscm/blob/master/notebooks/timeseriess.ipynb>`_.
```
Comment (znicholls) [inline openscm/timeseries_converter.py]:
```suggestion
<https://github.com/openclimatedata/openscm/blob/master/notebooks/timeseries.ipynb>`_.
```

File: openscm/timeseries_converter.py [hunk 4]
```diff
@@ -0,0 +1,364 @@
...
+    int_averages = [np.nan] * len(target_intervals[:-1])
+    for i, l in enumerate(target_intervals[:-1]):
+        u = target_intervals[i + 1]
+        y, _ = integrate.quad(continuous_representation, l, u)
+        int_averages[i] = y / (u - l)
+
+    return np.array(int_averages)
+
+
+def _calc_linearization_values(values: np.ndarray) -> np.ndarray:
```
Comment (znicholls) [inline openscm/timeseries_converter.py]: I think this can be deleted no? It doesn't get used anywhere and is a duplicate of calculate continuous?

Comment (znicholls) [reply openscm/timeseries_converter.py]: I lie ignore that comment. Perhaps we can rename somehow? `_calc_integral_preserving_linear_interpolation`?

File: openscm/timeseries_converter.py [hunk 5]
```diff
@@ -0,0 +1,364 @@
...
+        extrapolation_type
+            Extrapolation type
+        """
+        self._source = np.array(source_time_points, copy=True)
+        self._target = np.array(target_time_points, copy=True)
+        self._timeseries_type = timeseries_type
+        self._interpolation_type = interpolation_type
+        self._extrapolation_type = extrapolation_type
+
+        if source_time_points[0] > target_time_points[1]:
```
Comment (znicholls) [inline openscm/timeseries_converter.py]: this will become conditional on extrapolation_type in future right?

File: .gitignore [hunk 6]
```diff
@@ -17,3 +17,5 @@ dist/
 docs/_build
 venv/
 htmlcov/
+.pyre
+.pyre_configuration
```
Comment (znicholls) [inline .gitignore]: what is pyre?

File: openscm/timeseries_converter.py (hunk 5 above)
Comment (swillner) [reply openscm/timeseries_converter.py]: Yes, will add a comment

File: .gitignore (hunk 6 above)
Comment (swillner) [reply .gitignore]: A static type checker (http://pyre-check.org/), but I switched to Mypy (https://mypy.readthedocs.io/) now in #134 as it seems to be more complete and easier to configure

File: openscm/core.py (hunk 1 above)
Comment (swillner) [reply openscm/core.py]: We missed this probably

File: openscm/core.py [hunk 7]
```diff
@@ -214,17 +221,25 @@ def get_timeseries_view(
...
 
     def get_writable_timeseries_view(
         self,
         name: Tuple[str],
         region: Tuple[str],
         unit: str,
-        start_time: int,
-        period_length: int,
+        time_points: Tuple[int],
+        timeseries_type: ParameterType = ParameterType.AVERAGE_TIMESERIES,
```
Comment (znicholls) [inline openscm/core.py]: we should probably make the default point timeseries in future right? More intuitive?

Comment (swillner) [reply openscm/core.py]: yes. would also remove a default here for now if that's ok?

Comment (znicholls) [reply openscm/core.py]: good for me

**Comment by swillner** [comment]
@znicholls Cool. Can you merge this and then #134 please?

File: .pylintrc [hunk 8]
```diff
@@ -10,3 +10,9 @@ disable=
     too-many-locals,
     # handled by black
     format
+
+[SIMILARITIES]
+min-similarity-lines=6
```
Comment (znicholls) [inline .pylintrc]: why is this so high?

File: CHANGELOG.rst [hunk 9]
```diff
@@ -15,3 +15,4 @@ master
 - (`#44 <https://github.com/openclimatedata/openscm/pull/44>`_) Add timeframes module
 - (`#40 <https://github.com/openclimatedata/openscm/pull/40>`_) Add parameter handling in core module
 - (`#104 <https://github.com/openclimatedata/openscm/pull/104>`_) Define Adapter interface
+- (`#133 <https://github.com/openclimatedata/openscm/pull/104>`_) Cleanup and advance timeseries converter
```
Comment (znicholls) [inline CHANGELOG.rst]: @swillner can you add a CHANGELOG entry for #134 please?

File: .pylintrc (hunk 8 above)
Comment (swillner) [reply .pylintrc]: The default is 4, I think. Reverted back to the default and added comments to avoid the false positives.

File: CHANGELOG.rst (hunk 9 above)
Comment (swillner) [reply CHANGELOG.rst]: done

File: .pylintrc [hunk 10]
```diff
@@ -10,3 +10,8 @@ disable=
     too-many-locals,
     # handled by black
     format
+
+[SIMILARITIES]
+ignore-comments=no
+ignore-docstrings=yes
+ignore-imports=yes
```
Comment (znicholls) [inline .pylintrc]: doesn't duplicate docstrings and imports suggest something could be done better?

File: openscm/core.py [hunk 11]
```diff
@@ -256,20 +275,28 @@ def get_writable_timeseries_view(
...
-    def get_parameter_info(self, name: Tuple[str], region: Tuple[str]) -> ParameterInfo:
+        parameter.attempt_write(unit, timeseries_type, time_points)
+        return WritableTimeseriesView(
+            parameter,
+            unit,
+            time_points,
+            timeseries_type,
+            interpolation_type,
+            extrapolation_type,
+        )  # WritableTimeseriesView
```
Comment (znicholls) [inline openscm/core.py]:
```suggestion
        )
```

File: openscm/parameter_views.py [hunk 12]
```diff
@@ -166,14 +185,21 @@ def get_data_views_for_children_or_parameter(
...
+        self._unit_converter = UnitConverter(cast(str, parameter._info._unit), unit)
         self._timeseries_converter = TimeseriesConverter(
-            parameter._info._timeseries, timeseries
-        )
+            parameter._info._time_points,
+            time_points,
+            timeseries_type,
+            interpolation_type,
+            extrapolation_type,
+        )  # TimeseriesConverter
```
Comment (znicholls) [inline openscm/parameter_views.py]:
```suggestion
        )
```

File: openscm/regions.py [hunk 13]
```diff
@@ -2,7 +2,7 @@
 Handling of region information.
 """
 
-from typing import Dict, Tuple
+from typing import cast, Dict, Optional, Tuple
 from .errors import RegionAggregatedError
 from . import parameters  # pylint: disable=cyclic-import
```
Comment (znicholls) [inline openscm/regions.py]: why isn't the cyclic-import a problem?

File: openscm/timeseries_converter.py [hunk 14]
```diff
@@ -0,0 +1,355 @@
...
+        """
+        if len(values) < 3:
+            raise InsufficientDataError
+
+        if self._timeseries_type == ParameterType.AVERAGE_TIMESERIES:
+            return _calc_interval_averages(
+                self._calc_continuous_representation(source_time_points, values),
+                target_time_points,
+            )
+        # else
```
Comment (znicholls) [inline openscm/timeseries_converter.py]:
```suggestion

```

File: openscm/parameters.py [hunk 15]
```diff
@@ -4,9 +4,8 @@
 
 # pylint: disable=unused-import,protected-access
```
Comment (znicholls) [inline openscm/parameters.py]: why do we have unused imports?

File: openscm/units.py [hunk 16]
```diff
@@ -87,6 +87,10 @@
 """
 # pylint: disable=unused-import
```
Comment (znicholls) [inline openscm/units.py]: why do we have unused imports?

File: openscm/utils.py [hunk 17]
```diff
@@ -5,8 +5,10 @@
 from typing import Tuple, Union
 import warnings
 
+_ensure_input_is_tuple_calls: int = 0
```
Comment (znicholls) [inline openscm/utils.py]: why this change?

File: .pylintrc (hunk 10 above)
Comment (swillner) [reply .pylintrc]: In principle, yes, but especially imports can lead to false positives. Will change the docstrings, though.

File: openscm/core.py (hunk 11 above)
Comment (swillner) [reply openscm/core.py]: Unfortunately, pylint only checks for duplicate lines, not statements. These comments are needed to avoid false positives of duplicate code (see last commit).

File: openscm/regions.py (hunk 13 above)
Comment (swillner) [reply openscm/regions.py]: Cyclic imports are not necessarily a problem in Python (because of the `sys.modules` caching). This import, however, is necessary for Sphinx to find the correct types. Would be happy to do that in a better way, but have not found any.

File: openscm/parameters.py (hunk 15 above)
Comment (swillner) [reply openscm/parameters.py]: Sphinx (see comment above)

File: openscm/units.py (hunk 16 above)
Comment (swillner) [reply openscm/units.py]: Sphinx (see comment above)

File: openscm/utils.py (hunk 17 above)
Comment (swillner) [reply openscm/utils.py]: Pylint and Mypy were complaining. Both ways are equally good, I suppose.

**Comment by swillner** [comment]
In general, I think our Pylint configuration is still quite strict. I would prefer to keep it that way and only locally disable checks were absolutely necessary.

File: openscm/timeseries_converter.py [hunk 18]
```diff
@@ -0,0 +1,355 @@
...
+                )
+                res = interpolate.interp1d(
+                    linearization_points,
+                    linearization_values,
+                    fill_value="extrapolate"
+                    if self._extrapolation_type == ExtrapolationType.LINEAR
+                    else None,
+                )  # type: Callable[[float], float]
+                return res
+        # else
```
Comment (swillner) [inline openscm/timeseries_converter.py]:
```suggestion
```

File: .pylintrc (hunk 10 above)
Comment (swillner) [reply .pylintrc]: The default of 4 lines unfortunately leads to a lot of false postives for duplicate docstrings (especially because of the Sphinx-lines such as `Raises` etc.)

Comment (znicholls) [reply .pylintrc]: cool all good then

File: openscm/regions.py (hunk 13 above)
Comment (znicholls) [reply openscm/regions.py]: ah ok fair

File: openscm/utils.py (hunk 17 above)
Comment (znicholls) [reply openscm/utils.py]: ok. I have a particular hate for global variables which arose in another project so would prefer to avoid the global if possible. However if this is the only one I'll survive.

Comment (swillner) [reply openscm/utils.py]: Totally agree with you on globals ;) I have found a way to avoid that, just a second.

Comment (swillner) [reply openscm/utils.py]: Done

**Comment by swillner** [comment]
@znicholls Just these small changes ;) Would you merge this, please?

**Comment by znicholls** [comment]
wicked cheers @swillner
